packages/flutter/lib/src/rendering/selection.dart

```diff
@@ -100,17 +100,123 @@ abstract class SelectionHandler implements ValueListenable<SelectionGeometry> {
   SelectionResult dispatchSelectionEvent(SelectionEvent event);
 }
 
+/// Represents a range of content for a given [SelectedContent].
+///
+/// This class stores the information of an active selection under
+/// a [Selectable] or [SelectionHandler], including the start and
+/// end positions of a selection relative to the content, as well
+/// as an identifier to identify the owner of the content, any
+/// children selection ranges that may be contained within the active
+/// selection, and the content itself.
+///
+/// [SelectionArea] and [SelectableRegion] provide access to this
+/// information through the [SelectedContent] provided in their
+/// [SelectionArea.onSelectionChanged] and
+/// [SelectableRegion.onSelectionChanged] members.
+///
+/// {@tool dartpad}
+/// This example shows how to color red the active selection
+/// under a [SelectionArea] or [SelectableRegion].
+///
+/// ** See code in examples/api/lib/material/selection_area/selection_area.1.dart **
+/// {@end-tool}
+///
+/// {@tool dartpad}
+/// This example shows how to replace the active selection
+/// under a [SelectionArea] or [SelectableRegion] with a widget.
+///
+/// ** See code in examples/api/lib/material/selection_area/selection_area.2.dart **
+/// {@end-tool}
+///
+/// See also:
+///
+///   * [SelectedContent], which contains the [SelectedContentRange] for a given [Selectable] or [SelectionHandler].
+abstract class SelectedContentRange<T extends Object> {
+  /// Creates a range for the content of a [Selectable] or [SelectionHandler].
+  SelectedContentRange({
+    this.selectableId,
+    required this.content,
+    this.children,
+  });
+
+  /// The unique id for the [Selectable] that created the range.
+  final int? selectableId;
+
+  /// The content that contains the selection.
+  final T content;
```

Comment:
I know we talked offline about the type T extending Object, but trying to refresh myself. Is that because it's reasonable for it to be a widget? Instead of always an InlineSpan.

Reply:
I think it could be a widget, though that brings up a good point. I think if it were to be a widget we might need something like `contentLength` member. The widget could define its content length as `1` for example and the start: 0/end: 1. Without a `contentLength` start/end offsets are ambiguous for non text content.

Reply:
I think it's not always guaranteed that a List<SelectedContentRange<Object>> is homogenous, for example you can have a column of 4 text widgets followed by a selectable image, and that selectable may or not may not provide a `SelectedContentRange` that is of the same type.

Reply:
I think the use case of providing `void Function(List<SelectedContentRange<TextSpan>>? selections)` to `onSelectionChanged` can be accomplished by making `SelectionListener` generic as well as the `onSelectionChangedCallback`. What do you think about this?

Reply:
A generic `SelectionListener` doesn't seem possible since `getSelections` does not guarantee a homogenous type, this leads to a runtime error.